### PR TITLE
Support MapWithState

### PR DESCRIPTION
**Pull request by tawan0109**
* [Design doc ](https://microsoft.sharepoint.com/teams/sparkclr/_layouts/15/WopiFrame.aspx?sourcedoc={2be8dd3c-1ae3-40f5-8ab3-c440847c563c}&action=view)
* Unit tests will be appended to this PR after finish design and implementation review.
* Comment duplicated class CSharpMapWithStateDStream in CSharpDStream.scala for now.

### CODE REVIEW

**Comment by xiongrenyi** [comment]
Good work! Tao, also looks like review has lots to look at. Will do it when I get back from Spark Summit if conference schedule is busy.

Thanks,
Renyi.

**Comment by xiongrenyi** [comment]
A quick note:

I think CSharpMapWithStateRD should be implemented on C# side(?) like pyspark and SparkCLR always do?

You can handle (string, byte[], byte[]) on C# side instead of passing to JVM?

From: Renyi Xiong
Sent: Monday, February 15, 2016 8:38 PM
To: 'Microsoft/SparkCLR' <reply@reply.github.com>; Microsoft/SparkCLR <SparkCLR@noreply.github.com
Subject: RE: [SparkCLR] Support MapWithState (#303)

Good work! Tao, also looks like review has lots to look at. Will do it when I get back from Spark Summit if conference schedule is busy.

Thanks,
Renyi.

**Comment by xiongrenyi** [comment]
I mean operator’s logic should be done on C# side while JVM side does low level work like checkpoint and passing data bytes to/from C#.

Sent from my Windows 10 phone

From: Renyi Xiong<mailto:renyix@microsoft.com>
Sent: Tuesday, February 16, 2016 12:10 AM
To: Microsoft/SparkCLR<mailto:reply@reply.github.com>; Microsoft/SparkCLR<mailto:SparkCLR@noreply.github.com>
Subject: RE: [SparkCLR] Support MapWithState (#303)

A quick note:

I think CSharpMapWithStateRD should be implemented on C# side(?) like pyspark and SparkCLR always do?

You can handle (string, byte[], byte[]) on C# side instead of passing to JVM?

From: Renyi Xiong
Sent: Monday, February 15, 2016 8:38 PM
To: 'Microsoft/SparkCLR' <reply@reply.github.com>; Microsoft/SparkCLR <SparkCLR@noreply.github.com
Subject: RE: [SparkCLR] Support MapWithState (#303)

Good work! Tao, also looks like review has lots to look at. Will do it when I get back from Spark Summit if conference schedule is busy.

Thanks,
Renyi.

**Comment by tawan0109** [comment]
Thanks for reviewing.

yes, this operation can also be done on C# side. 
I have been working on another version of mapwithState in which operator logic is implemented in C# side. Will send a new PR and let's compare them side by side.

**Comment by tawan0109** [comment]
@xiongrenyi A CR has sent out in which operator logic is implemented on C# side.

**Comment by skaarthik** [comment]
Great job on the design doc, @tawan0109 Please post a copy at https://github.com/Microsoft/SparkCLR/tree/master/dev/designdocs for external access

File: csharp/Adapter/Microsoft.Spark.CSharp/Proxy/Ipc/SparkContextIpcProxy.cs [hunk 1]
```diff
@@ -20,7 +20,7 @@ internal class SparkContextIpcProxy : ISparkContextProxy
     {
         private JvmObjectReference jvmSparkContextReference;
         private JvmObjectReference jvmJavaContextReference;
-        private JvmObjectReference jvmAccumulatorReference;
+        internal JvmObjectReference jvmAccumulatorReference;
```
Comment (skaarthik) [inline csharp/Adapter/Microsoft.Spark.CSharp/Proxy/Ipc/SparkContextIpcProxy.cs]: instead of this I suggest creating an internal getter property like it is done for JvmSparkContextReference below

File: csharp/Adapter/Microsoft.Spark.CSharp/Interop/Ipc/SerDe.cs [hunk 2]
```diff
@@ -174,6 +179,20 @@ public static void Write(Stream s, byte value)
             totalWriteNum += 1;
         }
 
+        public static void WriteBytes(Stream s, byte[] value)
```
Comment (skaarthik) [inline csharp/Adapter/Microsoft.Spark.CSharp/Interop/Ipc/SerDe.cs]: This method could be confused with Write(Stream, byte[]) method below. Why is this necessary? Can we give a better name?

File: scala/src/main/org/apache/spark/api/csharp/SerDe.scala [hunk 3]
```diff
@@ -267,6 +267,10 @@ object SerDe {
     out.writeInt(value)
   }
 
+  def writeLong(out: DataOutputStream, value: Long): Unit = {
```
Comment (skaarthik) [inline scala/src/main/org/apache/spark/api/csharp/SerDe.scala]: Should we start using this method @ https://github.com/Microsoft/SparkCLR/pull/303/files#diff-d078e99a5fd52f304c58828ec704b3cfL201?

File: scala/src/main/org/apache/spark/streaming/api/csharp/CSharpMapWithStateDStream.scala [hunk 4]
```diff
@@ -0,0 +1,544 @@
+/*
```
Comment (skaarthik) [inline scala/src/main/org/apache/spark/streaming/api/csharp/CSharpMapWithStateDStream.scala]: Remember to replace this license header

File: csharp/Adapter/Microsoft.Spark.CSharp/Proxy/Ipc/SparkContextIpcProxy.cs (hunk 1 above)
Comment (tawan0109) [inline csharp/Adapter/Microsoft.Spark.CSharp/Proxy/Ipc/SparkContextIpcProxy.cs]: Sure

File: csharp/Adapter/Microsoft.Spark.CSharp/Interop/Ipc/SerDe.cs (hunk 2 above)
Comment (tawan0109) [inline csharp/Adapter/Microsoft.Spark.CSharp/Interop/Ipc/SerDe.cs]: How about "WriteBytesWithLength(Stream, byte[])"?

File: scala/src/main/org/apache/spark/api/csharp/SerDe.scala (hunk 3 above)
Comment (tawan0109) [inline scala/src/main/org/apache/spark/api/csharp/SerDe.scala]: Seems the above link is broken.

File: scala/src/main/org/apache/spark/streaming/api/csharp/CSharpMapWithStateDStream.scala (hunk 4 above)
Comment (tawan0109) [inline scala/src/main/org/apache/spark/streaming/api/csharp/CSharpMapWithStateDStream.scala]: Sure

**Comment by tawan0109** [comment]
Close this PR for now.
